File: app/src/main/java/com/juniperphoton/myersplash/adapter/MainListFragmentAdapter.java
package com.juniperphoton.myersplash.adapter;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;

import com.juniperphoton.myersplash.fragment.MainListFragment;
import com.juniperphoton.myersplash.model.UnsplashCategory;

public class MainListFragmentAdapter extends FragmentStatePagerAdapter {
    private MainListFragment.Callback mCallback;

    public MainListFragmentAdapter(MainListFragment.Callback callback, FragmentManager fm) {
        super(fm);
        mCallback = callback;
    }

    @Override
    public Fragment getItem(int position) {
        MainListFragment fragment = new MainListFragment();
        switch (position) {
            case 0:
                fragment.setCategory(UnsplashCategory.getFeaturedCategory(), mCallback);
                break;
            case 1:
                fragment.setCategory(UnsplashCategory.getNewCategory(), mCallback);
                break;
            case 2:
                fragment.setCategory(UnsplashCategory.getRandomCategory(), mCallback);
                break;
        }
        return fragment;
    }

    @Override
    public int getCount() {
        return 3;
    }
}

File: app/src/main/kotlin/com/juniperphoton/myersplash/adapter/MainListFragmentAdapter.kt
package com.juniperphoton.myersplash.adapter

import android.graphics.RectF
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.view.View
import android.view.ViewGroup
import com.juniperphoton.myersplash.App
import com.juniperphoton.myersplash.data.DaggerRepoComponent
import com.juniperphoton.myersplash.data.MainListPresenter
import com.juniperphoton.myersplash.data.RepoModule
import com.juniperphoton.myersplash.fragment.MainListFragment
import com.juniperphoton.myersplash.model.UnsplashImage

class MainListFragmentAdapter(private var callback: ((RectF, UnsplashImage, View) -> Unit)?,
                              fm: FragmentManager
) : FragmentStatePagerAdapter(fm) {
    override fun instantiateItem(container: ViewGroup?, position: Int): Any {
        val o = super.instantiateItem(container, position)
        if (o is MainListFragment) {
            inject(o, position)
        }
        return o
    }

    private fun inject(fragment: MainListFragment, position: Int) {
        val presenter = MainListPresenter()
        val component = DaggerRepoComponent.builder()
                .repoModule(RepoModule(App.instance, position, fragment)).build()
        component.inject(presenter)

        fragment.presenter = presenter
        fragment.onClickPhotoItem = callback
    }

    override fun getItem(position: Int): Fragment = MainListFragment()

    override fun getCount(): Int = 3
}